### DISCUSSION TITLE
Converting a pdf to an image with c under windows with msys2

### DISCUSSION THREAD

**Discussion by Oszimilian** [ImageMagick/ImageMagick#5613] [Help]
Hi together,
in my project, which i wrote in c under linux, i convert a pdf to an png. I build the library from source and it worked totally fine. Now i want to compile my project under windows with msys2. I installed the ImageMagick package in this environment and compiled it with success.
The problem is that this code doesn't work. It seems like that it works but it doesn't convert my pdf to png. 

This is my code:
```bash
void WCO_PNG_Set_ConvertToPNG(float resize)
{
    //resizing the pdf by an given factor and casting it to an int
    int x = (int)(resize * WCO_PDF_Ref()->pageSize[0]);
    int y = (int)(resize * WCO_PDF_Ref()->pageSize[1]);

    //converting both pages. The task and the solution sheet via a for loop
    for(int i = 0; i <= 1; i++)
    {

        MagickWand *newWand;

        MagickWandGenesis();

        newWand = NewMagickWand();

        MagickReadImage(newWand, WCO_PDF_Ref()->fileName[i] );

        MagickResizeImage(newWand,x,y,LanczosFilter);

        MagickContrastImage(newWand, 100);

        MagickSetImageCompressionQuality(newWand,100);

        MagickWriteImage(newWand, WCO_PNG_Ref()->fileName[i]);

        if(newWand) DestroyMagickWand(newWand);

        //displaying the converted pdf in the gtk window
        WCO_GUI_Show_Worksheet(i);
    }

    //Set the PNGShowFlag 
    //This is only done once. After that the flag is always set
    int i = _TRUE;
    WCO_PNG_Set(&i, set_showPNG);
}
```

If i print out the string behind `WCO_PDF_Ref()->fileName[i]` and 'WCO_PNG_Ref()->fileName[i]' i got this:
```bash
C:\msys64\home\user\Git\WCO\PDF\001.pdf
C:\msys64\home\user\Git\WCO\bin\png\task.png
```
These directories are exactly what i want.

Have anyone a idea what's wrong with this?

**Comment by snibgo** [comment]
Which of your IM operations fail?

**Reply by Oszimilian** [reply]
I cant say because the program does not break up at a specific point. I got no error-messages. I would estimate that the problem occurs at reading the input pdf but i don't really know.

**Comment by snibgo** [comment]
Functions like MagickReadImage() return MagickTrue is they succeed or MagickFalse if they fail.

**Reply by Oszimilian** [reply]
I checked it and it looks like `MagickReadImage()` crashes. The return value is equal to `MagickFalse`.

**Comment by snibgo** [comment]
Good, you have narrowed the problem down.

What is the value of "`i`"? So, does the problem occur on the first or second page?

Skipping over that question, what version of Ghostscript do you have?

**Reply by Oszimilian** [reply]
i noticed that i forgot to install ghost-script. But now i have version 9.53.3-3 installed. This leads also to an error. The value `i` is responsible for the directorie of the two pdf which i want to convert. The error occurs in both cases. 
I also checked the case of a corrupted pdf with a random pdf from the internet, but also no success with this.

**Comment by snibgo** [comment]
>  I got no error-messages.

I suggest you modify your program so you get error messages. See, for example, https://github.com/ImageMagick/ImageMagick/blob/main/www/source/wand.c .

**Reply by Oszimilian** [reply]
This is the output:
```bash
src/PNG.c WCO_PNG_Set_ConvertToPNG 115 NoDecodeDelegateForThisImageFormat 'PDF' @ error/constitute.c/ReadImage/741
```
Does that mean that the pdf-fileformate is  not excepted in general or only in my case, because under linux i have no problem with it.

**Reply by Oszimilian** [reply]
Moreover, if i type 'convert -version' in the msys2 terminal i got a list with the supported file-formats. PDF is not listed

**Reply by fmw42** [reply]
What version of Imagemagick?  

`convert -list format`

that should show PDF, check if rw (read and write)

PDF won't show from

`convert -version`

You need to have Ghostscript installed. It may or may not show as gs or gslib in the result from that command.

**Reply by Oszimilian** [reply]
Version is 7.1.0-39. 
'convert -list format' returns nothing, but 'convert -version' returns the delegates and there is gslib listed. 
But it can not be that pdf's dont work because if i convert a pdf manually from the terminal it works.

**Comment by snibgo** [comment]
> 'convert -list format' returns nothing,

Something is badly wrong.

As you have v7, I suggest you use "magick", not "convert".

**Reply by Oszimilian** [reply]
were is the difference between those two?

**Comment by fmw42** [comment]
On IM 7, use magick, not convert. But it sound like something may be wrong.  Please post what you get from 

magick -version

and

magick -list format

**Reply by Oszimilian** [reply]
Output of `magick -version`
```bash
Version: ImageMagick 7.1.0-39 Q16-HDRI x86_64 20175 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5)
Delegates (built-in): bzlib cairo djvu fftw flif fontconfig freetype gslib gvc heic jbig jng jp2 jpeg jxl lcms lqr ltdl lzma openexr pangocairo png ps raqm raw rsvg tiff webp wmf x
ml zlib
Compiler: gcc (12.2)

```

Output `magick -list format`
```bash
-bash: $'\302\203magick': command not found

```

**Reply by fmw42** [reply]
That is a problem. I do not understand the message.  I would suggest removing that version and reinstalling.

**Reply by Oszimilian** [reply]
I removed it and installed the official package from msys2. 
The output didn't changed. 
So is there another possibility to install ImageMagick with the headers-files under windows?

Yesterday i installed the library from source in the msys2 file-directory, but during the compilation of my project I encountered numerous problems. So could this be an approach to look after the issues and trying it this way?

**Reply by fmw42** [reply]
You should install a binary version not from source. The binary includes most of the relevant delegate libraries. The source requires you to install each delegate manually before installing Imagemagick.

**Reply by Oszimilian** [reply]
`magick -list format`works with the windows binary release. 
I installed it with the necessarily headers but how can i tell my compiler where all these stuff is located? 
I think the tutorial on the IM-webside explained it for Linux and not for windows.

**Reply by fmw42** [reply]
Sorry, I am not a Windows user. So cannot help with that. Perhaps one of the other uses may be help you more.

**Comment by snibgo** [comment]
I don't build with msys2, so I cannot help with that.

**Comment by Oszimilian** [comment]
OK, thanks for your help. 
I will try it with another approach.
